ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRewriteToDataSketchesRules.java

```diff
@@ -483,46 +489,44 @@ protected final SqlOperator getSqlOperator(String fnName) {
   }
 
   /**
-   * Rewrites {@code cume_dist() over (order by id)}.
+   * Provides a generic way to rewrite function into using an estimation based on CDF.
+   *
+   *  There are a few methods which could be supported this way: NTILE, CUME_DIST, RANK
    *
+   *  For example:
    *  <pre>
    *   SELECT id, CUME_DIST() OVER (ORDER BY id) FROM sketch_input;
-   *     ⇒ SELECT id, 1.0-ds_kll_cdf(ds, CAST(-id AS FLOAT) )[0]
+   *     ⇒ SELECT id, ds_kll_cdf(ds, CAST(id AS FLOAT) )[0]
    *       FROM sketch_input JOIN (
-   *         SELECT ds_kll_sketch(CAST(-id AS FLOAT)) AS ds FROM sketch_input
+   *         SELECT ds_kll_sketch(CAST(id AS FLOAT)) AS ds FROM sketch_input
    *       ) q;
    *  </pre>
    */
-  public static class CumeDistRewrite extends WindowingToProjectAggregateJoinProject {
+  public static abstract class AbstractRankBasedRewriteRule extends WindowingToProjectAggregateJoinProject {
 
-    public CumeDistRewrite(String sketchType) {
+    public AbstractRankBasedRewriteRule(String sketchType) {
       super(sketchType);
     }
 
-    @Override
-    protected VbuilderPAP buildProcessor(RelOptRuleCall call) {
-      return new VB(sketchType, call.builder());
-    }
+    protected static abstract class AbstractRankBasedRewriteBuilder extends VbuilderPAP {
 
-    private static class VB extends VbuilderPAP {
-
-      protected VB(String sketchClass, RelBuilder relBuilder) {
+      protected AbstractRankBasedRewriteBuilder(String sketchClass, RelBuilder relBuilder) {
         super(sketchClass, relBuilder);
       }
 
       @Override
-      boolean isApplicable(RexOver over) {
-        SqlAggFunction aggOp = over.getAggOperator();
+      final boolean isApplicable(RexOver over) {
         RexWindow window = over.getWindow();
-        if (aggOp.getName().equalsIgnoreCase("cume_dist") && window.orderKeys.size() == 1
-            && window.getLowerBound().isUnbounded() && window.getUpperBound().isUnbounded()) {
+        if (window.orderKeys.size() == 1
+            && window.getLowerBound().isUnbounded() && window.getUpperBound().isUnbounded()
```

Comment:
Fwiw, Hive does a complete mess about this. If there is no window frame, following the SQL standard:
- No order by spec -> rows, unbounded preceding, unbounded following
- Order by spec -> range, unbounded preceding, current row
This also aligns with most RDBMSs implementation.

We may need to revisit the logic in this `WindowingSpec` block at some point:
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/parse/WindowingSpec.java#L214
But I suspect there is compensation logic for that in execution code path and it may require some work.

Reply:
interesting; mostly the second :D
for the current functions (ntile,cume_dist) doesn't really make sense to set the window anything than unbounded (or at least I don't see a usecase for it)

I've tried this out for the below query:
```
select id,ntile(id) over (order by id rows between 1 preceding and 1 following) from sketch_input order by id nulls last;
```
* mysql: rejects it with an error that `ntile` doesn't support it
* psql: accepts and executes it without interpreting the preceding/following stuff correctly
* hive: stops with a semanticexception

Reply:
the logic to handle 
```
Order by spec -> range, unbounded preceding, current row
This also aligns with most RDBMSs implementation
```
I think at the time this rule fires it will see unbounded/unbounded...but it's very weird I'll open a separate ticket to invetigate that 

opened: HIVE-23775